content/1_docs/2_cookbook/3_templating/0_understanding-oop/recipe.txt

```diff
@@ -0,0 +1,462 @@
+
+```php "/book/ebook.php"
+<?php
+
+namespace Cookbook\Books;
+
+class EBook {
+    //… code as before
+}
+```
+
+If we open `index.php` after having made this change, we will be greeted with an error: "Fatal error: Uncaught Error: Class "Book" not found in …/index.php on line ….
+
+That is because the full name of the `Book` class is now `Cookbook\Books\Book` and `Cookbook\Books\Ebook` respectively.
+
+We can now either add the full classname when invoking our objects:
+
+```php "/book/index.php"
+$book = new Cookbook\Books\Book('The Hitchhiker\'s Guide to the Galaxy', 'Douglas Adams');
+```
+
+or we import the class…
+
+### Importing classes
+
+Instead of instantiating an object with its full class name, we can import the classes using the `use` keyword. Once that is done, we can continue using the short name when instatiation an object like before:
+
+
+```php "/book/index.php"
+<?php
+// make PHP display errors on screen
+ini_set('display_errors', 1);
+// require the book class
+require_once __DIR__ . '/book.php';
+require_once __DIR__ . '/ebook.php';
+
+use Cookbook\Books\Book;
+use Cookbook\Books\EBook;
+
+$book  = new Book(…);
+$ebook = new EBook(…);
+//…
+```
+
+## Static properties and methods
+
+Static properties and methods can be called directly without creating an instance of a class using the `new` keyword.
+
+You have probably come around such static methods when using Kirby. Same classes in the Toolkit, for example the `A`, Str` or 'Html` classes, have only static methods.
+
+That we don't have to explicitly instanciate an object is not the only difference, we also call these methods differently, for example:
+
+```php
+echo Str::upper('Kirby is my favorite CMS');
+```
+
+Instead of the arrow syntax used for methods that are called on an instance of a class like `$page->update()`, a double colon (or (link: https://en.wikipedia.org/wiki/Scope_resolution_operator text: "Paamayim Nekudotayim")) is used to access a static method or property of a class.
+
+Fetch another coffee or tea and then let's add a static method to our `Book` class as well, so that we can create the new object statically instead.
+
+```php "/book/book.php"
+<?php
+class Ebook extends Book
+{
+  
+  // ... other code
+
+  public static function create($title, $author) {
+    return new static($title, $author);
+  }
+
+}
+```
+
+This new static method does nothing but call the constructor of the class internally, so that we can instantiate a new object without the `new` keyword:
+
+```php "/book/index.php"
+$ebook2 = Ebook::create('The Pragmatic Programmer', 'David Thomas, Andrew Hunt');
+var_dump($ebook2);
+```
+
+One advantage of this approach is that we can now call other methods directly without having to wrap the object into parentheses.
+
+```php "/book/index.php"
+$title = (new Ebook('Form Design Patterns', 'Adam Silver'))->getTitle();
+```
+
+vs.
+
+```php "/book/index.php"
+$title = Ebook::create('Form Design Patterns', 'Adam Silver')->getTitle();
+```
+
```

Comment:
I'd add a note here that this pattern can be useful sometimes, but probably shouldn't be used all the time.